### ISSUE TITLE
Distributor hangs when restarting/killing Message Queueing service 

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.Distributor.Msmq#32]
When Message Queueing service is restarted then the distributor endpoint stops processing messages. This can cause outages in a production environment. 

When using one of the following distributor profiles

* NServiceBus.Master
* NServiceBus.MSMQMaster
* NServiceBus.Distributor
* NServiceBus.MSMQDistributor

## Expected behavior

When Message Queueing service is unavailable the distributor endpoint should shut down just as any other endpoint would do when not using either master profiles.

## Tested versions:

* NServiceBus v4.6.1 (Profile: NServiceBus.Master)
* NServiceBus v4.7.5 (Profile: NServiceBus.Master)
* NServiceBus v4.7.5 & NServiceBus.Distributor.MSMQ v4.4.8 (Profile: NServiceBus.MSMQMaster)
* NServiceBus v5.2.0 & NServiceBus.Distributor.MSMQ v5.0.0 (Profile: NServiceBus.MSMQMaster)
* NServiceBus v4.7.5 & NServiceBus.Distributor.MSMQ v4.4.8 (Profile: NServiceBus.MSMQMaster)
* NServiceBus v5.2.14 & NServiceBus.Distributor.MSMQ v5.0.4 (Profile: NServiceBus.MSMQDistributor)

## Partial workaround

When MSMQ gets restarted in production then the expected behavior should be that all services that depend on it be restarted too. This can be configured with `sc.exe`

NOTE: This will **not** restart your service when MSMQ is killed or crashes. Only when you restart MSMQ via `services.msc` or `sc.exe start/stop`

```
DESCRIPTION:
        Modifies a service entry in the registry and Service Database.
USAGE:
        sc <server> config [service name] <option1> <option2>...

OPTIONS:
NOTE: The option name includes the equal sign.
      A space is required between the equal sign and the value.
 type= <own|share|interact|kernel|filesys|rec|adapt|userown|usershare>
 start= <boot|system|auto|demand|disabled|delayed-auto>
 error= <normal|severe|critical|ignore>
 binPath= <BinaryPathName to the .exe file>
 group= <LoadOrderGroup>
 tag= <yes|no>
 depend= <Dependencies(separated by / (forward slash))>
 obj= <AccountName|ObjectName>
 DisplayName= <display name>
 password= <password>
```

In this scenario:

    sc.exe config [servicename] depend=MSMQ/MSDTC

This will make your service dependent on both MSMQ and MSDTC. Mind to use `/` and not `,` to separate services.

if you restart either of those services then your endpoint will be started too.

**Comment by andreasohlund** [comment]
I guess we're not invoking `.CriticalError` properly?

**Comment by janovesk** [comment]
My guess as well. Haven't had time to dig in yet though.

**Comment by janovesk** [comment]
Happens with Distributor Profiles as well.

**Comment by andreasohlund** [comment]
@ramonsmits so this means that the process is still running but the worker is no longer processing any messages until it is restarted?

**Comment by danielmarbach** [comment]
@ramonsmits @janovesk @andreasohlund Can we do the triage and make a decision?

**Comment by ramonsmits** [comment]
@danielmarbach Will test this Monday, from what I remember is that processing completely stops without any indication.

**Comment by ramonsmits** [comment]
@danielmarbach  @janovesk @andreasohlund 

Made a recording: https://youtu.be/gbZPxoAkr_M

The workers and sender initiate a shutdown but the distributor is not. Tested it with the *Master* and the *Distributor* roles.

What it means is that the distributor node keeps running while it does not process incoming messages.

**Comment by andreasohlund** [comment]
>What it means is that the distributor node keeps running while it does not process incoming messages.

Sounds like a patch to me? @janovesk ?

**Comment by danielmarbach** [comment]
Is this how we do triage ;)?

**Comment by ramonsmits** [comment]
Based on the triage WIP this should be a critical as this can cause outage in production.

**Comment by andreasohlund** [comment]
>Is this how we do triage ;)?

My bad should articulate better:

Since this will cause messages to not be processed in production when the msmq service is restarted I propose we consider this a patch since there is a high chance of users breaking their message SLA's.

The workaround will cover scenarios when the service does a proper restart but not when it's crashing. I don't think we can rely on users finding and applying our workaround so I'm still :+1 for a patch.

Do we need to get on a call or can we get agreement async? cc @Particular/nservicebus-maintainers

**Comment by ramonsmits** [comment]
Its a patch but is it a critical as then it requires backporting too.

**Comment by SzymonPobiega** [comment]
Is it a message loss scenario? If not then I believe it is not a critical.

**Comment by andreasohlund** [comment]
>Is it a message loss scenario? If not then I believe it is not a critical.

No not message loss. But imagine that you don't reboot the server for a few days => no messages processed during that time. While not message loss I would consider that an outage?

**Comment by SzymonPobiega** [comment]
OK, you're right @andreasohlund . According to the [process](https://github.com/Particular/PlatformDevelopment/blob/master/Guidelines/Patching.md) this needs to be patched and backported.

**Comment by ramonsmits** [comment]
@danielmarbach is there a possibility that this is resolved by 5.2.18/5.2.19 ?

**Comment by danielmarbach** [comment]
Yes but I havent verified

**Comment by danielmarbach** [comment]
@ramonsmits would you be able to verify it?

**Comment by ramonsmits** [comment]
@danielmarbach Yes, I can

**Comment by ramonsmits** [comment]
This is resolved by 5.2.18, I'll add this to the milestone.

**Comment by ramonsmits** [comment]
That doesn't work, different repo. Fixed by https://github.com/Particular/NServiceBus/releases/tag/5.2.18

V6: As the distributor isn't updated to V6 and this isn't an issue on the workers this also works for the distributor in legacy mode.
V4 and earlier: Those aren't supported anymore so don't require a fix as there is no data loss.
